gcc/d/d-decls.cc

```diff
@@ -101,7 +101,38 @@ VarDeclaration::toSymbol()
 	  return csym;
 	}
 
-      csym = new Symbol();
+      // Use only same Symbol for VarDeclarations with same mangle
+      if (!this->mangleOverride && isDataseg () && !(storage_class & STCextern))
+	{
+	  static hash_map<nofree_string_hash, Symbol *> duplicate_map(251);
+	  const char *vname = mangle(this);
+	  bool existed;
+
+	  Symbol **entry = &duplicate_map.get_or_insert(vname, &existed);
+	  if (!existed)
+	    {
+	      csym = new Symbol();
+	      *entry = csym;
+	    }
+	  else
+	    {
+	      csym = *entry;
+	      if (!csym->Sloc.equals(loc))
+		{
+		  error ("Found different variables with identical mangle: %s"
+		    " at %s and %s", vname, loc.toChars(), csym->Sloc.toChars());
```

Comment:
Without this check the glue layer would simply pass the same tree to the backend for both variables. A different option would be considering the loc in the hash for the hash table. But it seems this code is not required for variables at all:
```d
int x;
int x;
=> test.d(3): Error: variable test.x conflicts with variable test.x at test.d(1)
```
So I'll remove this.


The same check is not done for functions:
```d
void foo(){}
void foo(){}
=> test.d:(.text._D4test3fooFZv+0x8): multiple definition of `_D4test3fooFZv'
test.o:test.d:(.text._D4test3fooFZv+0x0): first defined here
```
And with this pull request and without the `loc` check, the function example would no longer generate a linker / assembler error. Instead it would simply hook up all references to foo to the first processed function. With the `loc` check it errors instead.